### DISCUSSION TITLE
Pulling parameters from a Cascade Aerotherm Plus Heat pump

### DISCUSSION THREAD

**Discussion by johnmacleod101** [john30/ebusd#1288] [General]
Hi, I have 2 Aerotherm Plus 12 KW heat pumps that i am trying to pull data from. I can successfully pull data and see 1 heat pump when I do a scan (ebusctl info) but I cant (I dont think) can see the second heat pump.
<img width="825" alt="Screenshot 2024-06-14 at 14 35 06" src="https://github.com/john30/ebusd/assets/80105356/ec13e223-7f1d-49bc-827f-6e49e79c96e0">
You can see on address 36 it just says scanning but does not return a result. Maybe this is second heat pump. Is there a way to resolve what that is on address 36?
Doing a full scan I get this and that address does not appear.
<img width="526" alt="Screenshot 2024-06-14 at 14 32 58" src="https://github.com/john30/ebusd/assets/80105356/5163259c-4c99-42f5-a39d-ced833add9a3">
Any help would be appreciated. I dont see any history of anyone with a cascade configuration using ebusd.
Thanks in advance.

**Comment by JonesGW-MGD** [comment]
So generally when you connect two similar devices (like boilers or heat pumps in cascade mode) which otherwise would try to secure the same ebus address, Vaillant solve this by having you install a VR32 bus coupler to give the second device a different address.

There is no way ebusd can 'guess' what device is behind a VR32 so there are no VR32 config files per say.

I would say your heat pump behind the VR32 is presenting as address 18. 

If you search this forum, you should be able to see how you can rename the heat pump file to this address.  I've posted a few responses on the forum.  (I am short of time currently but if you are struggling, re-post and I'll see about tracking these down).

You will likely also need to change the latency parameter as the VR32 does not play well with the official ebus spec

**Comment by JonesGW-MGD** [comment]
So, in a little bit more detail:

1. You're going to have to run the configuration files locally if you're not already doing this - essentially involves copying/cloning the repository of the configuration files to your local, and then changing the configpath on ebusd startup

2. You'll also need to change the latency parameter to 100000 as discussed in this post: https://github.com/john30/ebusd/issues/109

3. You then need to copy 08.hmu.csv and rename it 18.v32.csv.  (Obviously don't just rename it - you need a copy of the original 08.hmu.csv for the first heat pump as well!)

Basically, ebusd matches configuration files on start-up to the slave address of device (boiler, heat pump, controller etc.) and ID (and SW/HW versions if necessary) - see the wiki on auto config naming - https://github.com/john30/ebusd/wiki/4.7.-Automatic-configuration.  Because any device could sit behind a VR32 - a boiler, heat pump or an air recovery system as examples, there is no way ebusd can work this out so consequently there are no vr32 configuration files in the repository.  But since _you_ know it's your heat pump, creating your own VR32 file (18.v32.csv) with the message definitions for a heat pump will allow the ebusd auto config to pick this file up and all should work fine!

Here a few threads I've commented on where it seemed to work in a similar situation to yours:

https://github.com/john30/ebusd/discussions/1223#discussioncomment-8749250
https://github.com/john30/ebusd/discussions/1197#discussioncomment-8542697

**Reply by johnmacleod101** [reply]
Hi,
I changed latency, copied and renamed the 08.hmu.csv file and restarted.
Using ebusdctl info command I dont see any additional addresses  showing the additional HP and nothing using that config file.
I was already using the local configuration and changed the latency as per your info.
Will spend some more time tomorrow and see if I can make any more progress.
<img width="847" alt="Screenshot 2024-06-17 at 21 03 53" src="https://github.com/john30/ebusd/assets/80105356/b92d3ce3-751c-4b6e-94d8-d18f16dd904d">

Thanks

**Reply by JonesGW-MGD** [reply]
Hi John,

OK, that's puzzling, I would have expected that to pick up the new file....  

(I'll caveat any advice henceforth - I've not had to do this and I don't even have a VR32 myself so I don't want to mislead you in the wrong direction) 

I wonder whether you might learn something more if you temporarily started ebusd with debug logging and confined the scan to address 18?

I think it would look something like: `--loglevel=debug --scanconfig=18` added to your ebusd start-up.

Or maybe another check would be to temporarily isolate the new file in a different directory, and restart ebusd with:

`--checkconfig` (specifying the new --configpath and I imagine you'll still need the latency change too)

Good luck!

Gareth

**Reply by johnmacleod101** [reply]
Hi Gareth, Made a little progress. To summarize, If I just use setting as scanconfig=18 It finds the VR32 and associates the correct file. If I then do a ebusctl scan and then ebusctl info I get the following which is great picking up all the devices but now not HMU00.
<img width="807" alt="Screenshot 2024-06-18 at 20 17 08" src="https://github.com/john30/ebusd/assets/80105356/6431d70e-01c0-4462-b4fb-177d9699fe1b">
But, Then when I allow to scan for all devices (by removing =18) in configuration It does not find it again. Same as before. But with debug on and scanning through the log I see this.
2024-06-18 20:18:48.422 [main debug] found 1 matching scan config files from vaillant with prefix 18: done
2024-06-18 20:18:48.423 [main error] error reading scan config file vaillant/18.v32.csv for ID "v32", SW0117, HW9802: ERR: duplicate entry, vaillant/18.v32.csv:2: ERR: duplicate entry, duplicate ID
2024-06-18 20:18:48.423 [main error] scan config 18: ERR: read timeout
So I assume this means that there is some duplication/conflict with HMU00 device.

John

**Reply by JonesGW-MGD** [reply]
OK - well, looking back, when I've seen this working it was two devices contending for the same address like 08, but not the same unit - so like a HP and boiler, which would load different files - but you've got two HPs.  Will have a think tomorrow but looks like the parser gets to line 2 and complains because it's identical to line 2 in the HMU file.  (line 1 is a comment so makes sense).  I had probably assumed that the address (ZZ) would differentiate it but maybe not.

I suspect you can simply make the files different by setting [CIRCUIT] - i.e. 'HP1' and 'HP2' in the respective files - again nothing I've done before but feels like some progress....!

**Reply by JonesGW-MGD** [reply]
So now definitely into realms of something I've not done and can't easily experiment with either.  

My seemingly incorrect assumption was that renaming the file would make each row in the file unique w.r.t to the other file since section 4.2 (end) of the wiki suggested to me that certain elements from the filename would be defaulted into rows if they were otherwise missing.  So row 2 in 18.v32.csv for example:

`*r,,,,,,"B511",,,,,,,,,,,,,,,,,,,,,,,,,`

would become:

`*r,,,,,18,"B511",,,,,,,,,,,,,,,,,,,,,,,,,`

when parsed since ZZ or the slave address is position/column 6.  This would make the row unique but of course position 8 ID is still blank and so would be a duplicate....

I think I'd be inclined to play around about with the files to see if I could understand the behaviour - so maybe change line 2 in 18.v32.csv to:

`*r,HP2,,,,18,"B511",,,,,,,,,,,,,,,,,,,,,,,,,`

In doing that does the log now show an error at a different line or is it still - vaillant/18.v32.csv:**2** << ?


If that changed the error position then it might be as simple as adding a CIRCUIT name to both files - again 4.2 in the wiki suggests you ought to be able to do that by renaming the files:

08.hmu.HP1.csv and 18.v32.HP2.csv or you could just manually amend each line so the second position had the respective circuit name.  (If you do this, you should be able to just amend the 'default lines' - that is, rows that start *r

I'm not sure what you can do if each ID column needs to unique instead of each row - I can't see how you could do that.

Beyond that I'd just be thrashing around a bit and re-reading the wiki to see whether it made any more sense!

Hopefully somebody else might come to the rescue, but as you observe, there is very little mention on any thread of a true cascade configuration where the heat source are the same unit, and only a few references to V32 or VR32s anyway and where typically they are different heat sources behind the bus coupler.

**Reply by johnmacleod101** [reply]
Hi Gareth,
Thanks for the suggestions for direction. If nothing else it has eliminated a few things. It seems no mater what I do I cannot get both heat pumps to attach to a config file at same time unless I remark (#) out every line in the file 18.v32.HP2.CSV and then I get both pumps attaching to their assigned config file.
<img width="1064" alt="Screenshot 2024-06-19 at 19 12 09" src="https://github.com/john30/ebusd/assets/80105356/73a4c617-9b66-43df-8f68-5795ff02a946">
I did as you suggested changing the circuit name but still it would not allow both devices to attach to a config file. The message is indicating that the ID in the file is finding a duplicate which I assume is because 18 is the second heat pump being scanned and is finding duplicate ID's that are already assigned I assume.
I then went back to just enabling 18 scanconfig=18 just o be sure this was actually the heat pump incase I was totally going in wrong direction.
Below is the MQTT scan for the heat pump and can confirm this is the other heat pump.
I will need to do some reading on these configuration files to see what the next steps are.
<img width="755" alt="Screenshot 2024-06-19 at 20 40 08" src="https://github.com/john30/ebusd/assets/80105356/b39962d9-b3b4-43d0-8c57-e30859ed8f9a">
The heat pump cascade that I have is not unusual and I am sure someone else will come across this issue also unless I am doinf something totally wrong.
John

**Reply by JonesGW-MGD** [reply]
How frustrating.

If you are minded, try these files - all I have done is specifically add circuit names and destination addresses explicitly on the *r lines, so may be nothing more than you've done via the file names, but perhaps if the defaults aren't getting pulled from the file name as suggested in 4.2.....

The wiki specifically says (4.1) :

_[ID]
Further ID bytes in hex placed in the first master data bytes (e.g. "040F01"). Together with the primary/secondary command bytes, the destination address, and the optional source address, this forms the unique message ID._

To me, this suggests the 'ID' at line 2, position 8 would be added to the now explicitly stated destination address in position 6, creating the '_unique message ID_'?

(The term 'ID' is confusing - whether it applies to the ID column in the file or the "ID" or IDENT (see 4.7) in the file name 

Assuming the 'ID' that is referred to in the error actually means this 'ID" in position 8 instead of ID meaning the ID or [IDENT] or "V32", (which is how ID is described in section 4.7) - I'm not sure how it now can be duplicated?

`2024-06-18 20:18:48.423 [main error] error reading scan config file vaillant/18.v32.csv for ID "v32", SW0117, HW9802: ERR: duplicate entry, vaillant/18.v32.csv:2: ERR: duplicate entry, duplicate ID`

[08.hmu.csv](https://github.com/user-attachments/files/15912783/08.hmu.csv)

[18.v32.csv](https://github.com/user-attachments/files/15912786/18.v32.csv)

Do you get exactly the same error with these? (That is, it's indicating the error is at line 2 or :2: in the error code)

Otherwise, totally out of ideas, sorry.

**Reply by johnmacleod101** [reply]
Hi,
Well looks like you are onto something. With these 2 files both devices are mapped to config files
<img width="785" alt="Screenshot 2024-06-20 at 20 03 48" src="https://github.com/john30/ebusd/assets/80105356/4cd50a68-bf92-4845-9401-def0a0872414">
But as i am sure you expected there are still duplicates
For address 08:

```
> 2024-06-20 20:02:20.133 [main notice] read scan config file vaillant/08.hmu.csv for ID "hmu00", SW0607, HW5103
> 2024-06-20 20:02:20.136 [main error] error executing instructions: ERR: duplicate entry, last error: error including "vaillant/hcmode.inc" for "08": ERR: duplicate entry vaillant/hcmode.inc:22: ERR: duplicate entry, duplicate ID, included "vaillant/errors.inc" for "08"
> 2024-06-20 20:02:20.136 [main notice] found messages: 71 (0 conditional on 0 conditions, 0 poll, 13 update)
> 2024-06-20 20:02:20.136 [bus info] scan 08 cmd: 3108b5090124
```

So looks like it is now getting to the lines with !include. 


For address 18

```
2024-06-20 20:02:26.076 [main notice] read scan config file vaillant/18.v32.csv for ID "v32", SW0117, HW9802
2024-06-20 20:02:26.077 [main error] error executing instructions: ERR: duplicate entry, last error: error including "vaillant/hcmode.inc" for "18": ERR: duplicate entry vaillant/hcmode.inc:22: ERR: duplicate entry, duplicate ID, included "vaillant/errors.inc" for "18"
2024-06-20 20:02:26.077 [main notice] found messages: 588 (0 conditional on 0 conditions, 0 poll, 14 update)
2024-06-20 20:02:26.077 [main info] scan config 18 message received
```

So for the include files should I create 2 new errors.in and 2 new hcmode.inc files and customize them and define the new circuits for each heat pump?
I have been busy with other jobs today and only managed to drop the files in and try but need to get my head around what you changed in the files to get this far.
Really appreciate the help.
Kind regards
John

**Reply by JonesGW-MGD** [reply]
Hi John - small steps! 

I'll have a think in the am but yes, you could maybe customise each hcmode and errors file to have the [CIRCUIT] and [ZZ] specifically in each and rename and re-point but I'm not totally sure how they work - might be you can just omit the includes from one of the files.

But I DNU why it's suggesting an error at line 22 in hcmode.inc - doesn't seem to exist? So I maybe don't understand the error line semantics myself.  But I've had a whisky already tonight so maybe tomorrow...!

Notwithstanding the errors - do you seem more data on a scan or via MQTT?

Gareth

**Reply by JonesGW-MGD** [reply]
for what it's worth...

I removed the include statement for hcmode.inc and included the hcmode messages into each, adding the circuit name and the slave address as I had done for the main messages.  

I didn't do this for the errors.inc file - I have two files - my boiler message file and my SensoCOMFORT controller message file that both include the same errors.inc so can't see that that is causing the issue.

Not very elegant and probably not the best thing to do but may help with further troubleshooting?

[08.hmu.csv](https://github.com/user-attachments/files/15925328/08.hmu.csv)
[18.v32.csv](https://github.com/user-attachments/files/15925333/18.v32.csv)

**Reply by johnmacleod101** [reply]
Hi Gareth,
A quick update. I replaced the 08.hmu.csv and the 18.v32.csv files with the files you created and I can see the data from both pumps now and no error when loading the files. Really appreciate your help. I wont have time for a couple of days but going to try and summarise so that if someone else has this it will be easy to navigate to a solution.

**Reply by JonesGW-MGD** [reply]
Fingers crossed then!  Hopefully you get some good data.  I've been very happy to help - When I got into ebusd in Feb after having replaced my night storage heaters with a LPG gas combi late last year (rural, old property not really suitable for HP and no chance of solar on the straw roof..!), I felt like a kid in a toy shop with the stuff I got out (and am still logging).  I found the discussion forums invaluable especially since I'm technically minded but not a 'techie' so whatever small contribution I can make in return seems fair!  Good luck!  Gareth

**Comment by JonesGW-MGD** [comment]
Oh - and it's the ebus adapter you have on 31/36 (Looks like you've got a nice setup btw!)

**Reply by johnmacleod101** [reply]
Thanks for all the information on how to set up to read data from the second heatpump. You you are correct there is a bus coupler for the 2 heat pumps.
I will give this a try tomorrow and let you know how I get on.
Really appreciate support and direction in how to get this working.
Kind regards
John
